core/consensus/transport.go

```diff
@@ -0,0 +1,188 @@
+	if !ok {
+		return nil, errors.New("unknown value")
+	}
+
+	return data, nil
+}
+
+// Broadcast creates a msg and sends it to all peers (including self).
+func (t *transport) Broadcast(ctx context.Context, typ qbft.MsgType, duty core.Duty,
+	peerIdx int64, round int64, valueHash [32]byte, pr int64, pvHash [32]byte,
+	justification []qbft.Msg[core.Duty, [32]byte],
+) error {
+	// Get the values by their hashes.
+	value, err := t.getValue(valueHash)
+	if err != nil {
+		return err
+	}
+	pv, err := t.getValue(pvHash)
+	if err != nil {
+		return err
+	}
+
+	// Make the message
+	msg, err := createMsg(typ, duty, peerIdx, round, value, pr, pv, justification)
+	if err != nil {
+		return err
+	}
+
+	// First send to self.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case t.recvBuffer <- msg:
+	}
+
+	for i, p := range t.component.peers {
+		if int64(i) == t.component.peerIdx {
+			// Do not broadcast to self
+			continue
+		}
+
+		err := send(ctx, t.component.tcpNode, p.ID, msg.ToConsensusMsg())
+		if err != nil {
+			log.Warn(ctx, "Failed sending message", err)
+		}
+	}
+
+	return nil
+}
+
+// ProcessReceives processes received messages from the outer buffer until the context is closed.
+func (t *transport) ProcessReceives(ctx context.Context, outerBuffer chan msg) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case msg := <-outerBuffer:
+			if err := validateMsg(msg); err != nil {
+				log.Warn(ctx, "Dropping invalid message", err)
+				continue
+			}
+			t.setValues(msg)
+
+			select {
+			case <-ctx.Done():
+				return
+			case t.recvBuffer <- msg:
+			}
+		}
+	}
+}
+
+// createMsg returns a new message by converting the inputs into a protobuf
+// and wrapping that in a msg type.
+func createMsg(typ qbft.MsgType, duty core.Duty,
+	peerIdx int64, round int64, value *pbv1.UnsignedDataSet,
+	pr int64, pv *pbv1.UnsignedDataSet,
+	justification []qbft.Msg[core.Duty, [32]byte],
+) (msg, error) {
+	pbMsg := &pbv1.QBFTMsg{
+		Type:          int64(typ),
+		Duty:          core.DutyToProto(duty),
+		PeerIdx:       peerIdx,
+		Round:         round,
+		Value:         value,
+		PreparedRound: pr,
+		PreparedValue: pv,
+	}
+
+	// TODO(corver): Sign message.
+
+	// Transform justifications into protobufs
+	var justMsgs []*pbv1.QBFTMsg
+	for _, j := range justification {
+		impl, ok := j.(msg)
+		if !ok {
+			return msg{}, errors.New("invalid justificationProtos")
+		}
+		justMsgs = append(justMsgs, impl.msg) // Note nested justifications are ignored.
+	}
+
+	return newMsg(pbMsg, justMsgs)
+}
+
+// validateMsg returns an error if the message is invalid.
+func validateMsg(_ msg) error {
+	// TODO(corver): implement (incl signature verification).
+	return nil
+}
+
+// send sends the protobuf message to the peer.
+func send(ctx context.Context, tcpNode host.Host, id peer.ID, pb proto.Message) error {
+	s, err := tcpNode.NewStream(ctx, id, protocolID)
+	if err != nil {
+		return errors.Wrap(err, "new stream")
+	}
+	defer s.Close()
```

Comment:
There's no scenario where s can be non-nil when err is non-nil, right? Otherwise we could have a memory leak from returning before deferring the close.